diff -ruN DoubleLinkedList.sol DoubleLinkedList.sol
--- DoubleLinkedList.sol	2024-07-12 09:49:49.899261374 +0200
+++ DoubleLinkedList.sol	2024-07-12 09:52:16.333105454 +0200
@@ -16,6 +16,8 @@
 
     struct List {
         mapping(address => Account) accounts;
+        address insertedBefore; // HARNESS: address of the account before which the account was inserted at last insertion.
+        address insertedAfter; // HARNESS: address of the account after which the account was inserted at last insertion.
     }
 
     /* ERRORS */
@@ -90,21 +92,20 @@
     /// @param list The list to search in.
     /// @param id The address of the account.
     /// @param value The value of the account.
-    /// @param maxIterations The max number of iterations.
-    function insertSorted(List storage list, address id, uint256 value, uint256 maxIterations) internal {
+    function insertSorted(List storage list, address id, uint256 value) internal {
         if (value == 0) revert ValueIsZero();
         if (id == address(0)) revert AddressIsZero();
         if (list.accounts[id].value != 0) revert AccountAlreadyInserted();
 
         address next = list.accounts[address(0)].next; // `id` will be inserted before `next`.
 
-        uint256 numberOfIterations;
-        for (; numberOfIterations < maxIterations; numberOfIterations++) {
+        for (;;) {
             if (next == address(0) || list.accounts[next].value < value) break;
+            list.insertedAfter = next; // HARNESS
             next = list.accounts[next].next;
         }
 
-        if (numberOfIterations == maxIterations) next = address(0);
+        list.insertedBefore = next; // HARNESS
 
         address prev = list.accounts[next].prev;
         list.accounts[id] = Account(prev, next, value);
