diff -ruN DoubleLinkedList.sol DoubleLinkedList.sol
--- DoubleLinkedList.sol
+++ DoubleLinkedList.sol
@@ -16,6 +16,8 @@
 
     struct List {
         mapping(address => Account) accounts;
+        address insertedBefore; // HARNESS: address of the account before which the account was inserted at last insertion.
+        address insertedAfter; // HARNESS: address of the account after which the account was inserted at last insertion.
     }
 
     /* ERRORS */
@@ -101,10 +103,12 @@
         uint256 numberOfIterations;
         for (; numberOfIterations < maxIterations; numberOfIterations++) {
             if (next == address(0) || list.accounts[next].value < value) break;
+            list.insertedAfter = next; // HARNESS
             next = getNext(list, next);
         }
 
         if (numberOfIterations == maxIterations) next = address(0);
+        list.insertedBefore = next; // HARNESS
 
         address prev = getPrev(list, next);
         list.accounts[id] = Account(prev, next, value);
