diff -ruN DLL.sol DLL.sol
--- DLL.sol	1970-01-01 01:00:00.000000000 +0100
+++ DLL.sol	2022-09-29 18:07:48.707303249 +0200
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+import "./DoubleLinkedList.sol";
+
+contract DLL {
+    using DoubleLinkedList for DoubleLinkedList.List;
+
+    DoubleLinkedList.List public dll;
+
+    uint256 internal dummy_state_variable;
+
+    function dummy_state_modifying_function() public {
+        // to fix a CVL error when only one function is accessible
+        dummy_state_variable = 1;
+    }
+
+    function getValueOf(address _id) public view returns (uint256) {
+        return dll.getValueOf(_id);
+    }
+
+    function getHead() public view returns (address) {
+        return dll.head;
+    }
+
+    function getTail() public view returns (address) {
+        return dll.getTail();
+    }
+
+    function getNext(address _id) public view returns (address) {
+        return dll.getNext(_id);
+    }
+
+    function getPrev(address _id) public view returns (address) {
+        return dll.getPrev(_id);
+    }
+
+    function remove(address _id) public {
+        dll.remove(_id);
+    }
+
+    function insertSorted(address _id, uint256 _value) public {
+        dll.insertSorted(_id, _value);
+    }
+
+    function getInsertAfter() public view returns (address) {
+        return dll.insertAfter;
+    }
+
+    function getInsertBefore() public view returns (address) {
+        return dll.insertBefore;
+    }
+
+    function getLength() public view returns (uint256) {
+        uint256 len;
+        for (address current = dll.head; current != address(0); len++) current = getNext(current);
+        return len;
+    }
+
+    function isForwardLinkedBetween(
+        address _start,
+        address _end,
+        uint256 _maxIterations
+    ) public view returns (bool) {
+        if (_start == address(0) && _end == address(0)) return true;
+        for (; _maxIterations > 0; _maxIterations--) {
+            if (_start == _end) return true;
+            _start = dll.getNext(_start);
+        }
+        return false;
+    }
+
+    function isDecrSortedFrom(address _start, uint256 _maxIterations) public view returns (bool) {
+        for (; _maxIterations > 0; _maxIterations--) {
+            address next = dll.getNext(_start);
+            if (next == address(0)) return true;
+            if (getValueOf(_start) < getValueOf(next)) return false;
+            _start = dll.getNext(_start);
+        }
+        return true;
+    }
+}
diff -ruN DoubleLinkedList.sol DoubleLinkedList.sol
--- DoubleLinkedList.sol	2022-05-31 14:28:35.619981325 +0200
+++ DoubleLinkedList.sol	2022-09-29 18:04:49.699311765 +0200
@@ -14,6 +14,8 @@
         mapping(address => Account) accounts;
         address head;
         address tail;
+        address insertBefore; // HARNESS
+        address insertAfter; // HARNESS
     }
 
     /// ERRORS ///
@@ -89,33 +91,27 @@
     /// @param _list The list to search in.
     /// @param _id The address of the account.
     /// @param _value The value of the account.
-    /// @param _maxIterations The max number of iterations.
     function insertSorted(
         List storage _list,
         address _id,
-        uint256 _value,
-        uint256 _maxIterations
+        uint256 _value
     ) internal {
         if (_value == 0) revert ValueIsZero();
         if (_id == address(0)) revert AddressIsZero();
         if (_list.accounts[_id].value != 0) revert AccountAlreadyInserted();
 
-        uint256 numberOfIterations;
-        address next = _list.head; // If not added at the end of the list `_id` will be inserted before `next`.
+        _list.insertAfter = address(0);
+        address next = _list.head; // `_id` will be inserted before `next`.
 
-        while (
-            numberOfIterations < _maxIterations &&
-            next != _list.tail &&
-            _list.accounts[next].value >= _value
-        ) {
+        while (next != address(0) && _list.accounts[next].value >= _value) {
+            _list.insertAfter = next;
             next = _list.accounts[next].next;
-            unchecked {
-                ++numberOfIterations;
-            }
         }
 
+        _list.insertBefore = next;
+
         // Account is not the new tail.
-        if (next != address(0) && _list.accounts[next].value < _value) {
+        if (next != address(0)) {
             // Account is the new head.
             if (next == _list.head) {
                 _list.accounts[_id] = Account(address(0), next, _value);
diff -ruN .gitignore .gitignore
--- .gitignore	1970-01-01 01:00:00.000000000 +0100
+++ .gitignore	2022-09-29 18:04:49.699311765 +0200
@@ -0,0 +1,2 @@
+*
+!.gitignore
diff -ruN test/TestRedBlackBinaryTreeOptimized.sol test/TestRedBlackBinaryTreeOptimized.sol
--- test/TestRedBlackBinaryTreeOptimized.sol	2022-05-31 14:27:08.709981906 +0200
+++ test/TestRedBlackBinaryTreeOptimized.sol	1970-01-01 01:00:00.000000000 +0100
@@ -1,42 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.8.0;
-
-import "../RedBlackBinaryTreeOptimized.sol";
-
-contract TestRedBlackBinaryTreeOptimized {
-    using RedBlackBinaryTreeOptimized for RedBlackBinaryTreeOptimized.Tree;
-
-    RedBlackBinaryTreeOptimized.Tree public tree;
-
-    function insert(address _id, uint256 _value) external {
-        tree.insert(_id, _value);
-    }
-
-    function remove(address _id) external {
-        tree.remove(_id);
-    }
-
-    function keyExists(address _id) external view {
-        tree.keyExists(_id);
-    }
-
-    function last() external view {
-        tree.last();
-    }
-
-    function returnLast() external view returns (address) {
-        return tree.last();
-    }
-
-    function first() external view returns (address) {
-        return tree.first();
-    }
-
-    function next(address _key) external view returns (address) {
-        return tree.next(_key);
-    }
-
-    function returnKeyToValue(address _key) external view returns (uint256) {
-        return tree.keyToValue[_key];
-    }
-}
diff -ruN test/TestRedBlackBinaryTree.sol test/TestRedBlackBinaryTree.sol
--- test/TestRedBlackBinaryTree.sol	2022-05-31 14:28:35.619981325 +0200
+++ test/TestRedBlackBinaryTree.sol	1970-01-01 01:00:00.000000000 +0100
@@ -1,50 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.8.0;
-
-import "../RedBlackBinaryTree.sol";
-
-contract TestRedBlackBinaryTree {
-    using RedBlackBinaryTree for RedBlackBinaryTree.Tree;
-
-    RedBlackBinaryTree.Tree public tree;
-
-    function insert(address _id, uint256 _value) external {
-        tree.insert(_id, _value);
-    }
-
-    function remove(address _id) external {
-        tree.remove(_id);
-    }
-
-    function keyExists(address _id) external view {
-        tree.keyExists(_id);
-    }
-
-    function last() external view {
-        tree.last();
-    }
-
-    function returnLast() external view returns (uint256) {
-        return tree.last();
-    }
-
-    function first() external view returns (uint256) {
-        return tree.first();
-    }
-
-    function next(uint256 _value) external view returns (uint256) {
-        return tree.next(_value);
-    }
-
-    function returnKeyToValue(address _key) external view returns (uint256) {
-        return tree.keyToValue[_key];
-    }
-
-    function valueKeyAtIndex(uint256 _value, uint256 _index) external view returns (address) {
-        return tree.valueKeyAtIndex(_value, _index);
-    }
-
-    function getNumberOfKeysAtValue(uint256 _value) external view returns (uint256) {
-        return tree.getNumberOfKeysAtValue(_value);
-    }
-}
