diff -ruN DoubleLinkedList.sol DoubleLinkedList.sol
--- DoubleLinkedList.sol	2024-07-12 09:49:49.899261374 +0200
+++ DoubleLinkedList.sol	2024-07-12 09:53:11.964303440 +0200
@@ -16,6 +16,8 @@
 
     struct List {
         mapping(address => Account) accounts;
+        address insertedBefore; // HARNESS: address of the account before which the account was inserted at last insertion.
+        address insertedAfter; // HARNESS: address of the account after which the account was inserted at last insertion.
     }
 
     /* ERRORS */
@@ -101,10 +103,12 @@
         uint256 numberOfIterations;
         for (; numberOfIterations < maxIterations; numberOfIterations++) {
             if (next == address(0) || list.accounts[next].value < value) break;
+            list.insertedAfter = next; // HARNESS
             next = list.accounts[next].next;
         }
 
         if (numberOfIterations == maxIterations) next = address(0);
+        list.insertedBefore = next; // HARNESS
 
         address prev = list.accounts[next].prev;
         list.accounts[id] = Account(prev, next, value);
